ops/hookcmds/_action.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2025 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import json
+from typing import (
+    Any,
+    Mapping,
+    MutableMapping,
+    cast,
+    overload,
+)
+
+from ._utils import run
+
+
+def format_result_dict(
+    input: Mapping[str, Any],
+    parent_key: str | None = None,
+    output: dict[str, str] | None = None,
+) -> dict[str, str]:
+    """Turn a nested dictionary into a flattened dictionary, using '.' as a key separator.
+
+    This is used to allow nested dictionaries to be translated into the dotted
+    format required by the Juju `action-set` hook command in order to set nested
+    data on an action.
+
+    Example::
+
+        >>> test_dict = {'a': {'b': 1, 'c': 2}}
+        >>> format_result_dict(test_dict)
+        {'a.b': 1, 'a.c': 2}
+
+    Arguments:
+        input: The dictionary to flatten
+        parent_key: The string to prepend to dictionary's keys
+        output: The current dictionary to be returned, which may or may not yet
+            be completely flat
+
+    Returns:
+        A flattened dictionary
+
+    Raises:
+        ValueError: if the dict is passed with a mix of dotted/non-dotted keys
+            that expand out to result in duplicate keys. For example:
+            ``{'a': {'b': 1}, 'a.b': 2}``.
+    """
```

Comment:
This is not an exported function, so I wonder if a large, detailed doc string is called for.


Minor: it would be nice to document what happens to other types of values, e.g.:

`{'foo': 2**64}`, effectively coerced to a string at exec() time, using the Python rules
`...: [1,2,3]`, likewise
`{'foo.0': 42, 'foo.1': 42}`, Juju will later reassemble that to `{foo: {0: 42, 1: 42}}`
`{'return-code': "whatever"}`, lost, Juju overwrites with `return-code: 0`

That being said, I'm not sure where to document this.